README.md
InfiniumDiffMetMotR
===================
Version: 1.5.5

Description: This is a R package to analyze transcription factor binding motif enrichment at differentially methylated regions for Infinium Methylation BeadChip (Illumina).  

Last Update: 2020-08-24  

Depends: R (>= 2.10), Biobase (>= 2.5.5)  

Author: Takahiro Suzuki  

Updated by: takahiro.suzuki.aa@riken.jp

## Install
---
## 1. Install of devtools
devtools needs to install packages from github
```r
install.packages("devtools")
```
## 2. Install of InfiniumDiffMetMotR from github
```r
devtools::install_github("takahirosuzuki1980/InfiniumDiffMetMotR")
```
## Examples of usage
***
### ***TF binding Motif overrepresentation analysis***

### 1. Load of InfiniumDiffMetMotR
```r
library("InfiniumDiffMetMotR")
```

### 2. Normalization  
#### - row signal text data (generated by GenomeStudio)
```r
selDataMatrix <- lumiMethyNorm(fileName = "TableControl.txt", inputtype = "signal", sample_names = c("sample1", "sample2"))
```
- ***filename: unnourmalized-no background***
correction full infinium methylation array data.***  
The header of the data should includes the following items as the following formats;  
**TargetID    Sample1.Signal_A    Sample1.Signal_B    Sample1.Detection Pval**  
\- TargetID: Illumina probe ID  
\- Sample1.Signal_A: Signal of Unmethylated probe of sample 1  
\- Sample1.Signal_B: Signal of Methylated probe of sample 1  
\- Sample1.Detection Pval: detection P value of sample 1  
- ***inputtype: "signal" for a row signal text data***  
- ***sample_names: a vector of sample names***

#### - idat files
```r
selDataMatrix <- lumiMethyNorm(idatpath=getwd(), inputtype = "idat", sample_names = c("sample1", "sample2"))
```
- ***idatpath: A directory of idat files***  
a set of idat file name should be
**barcode_location_Grn.idat & barcode_location_Red.idat or GEOID_barcode_location_Grn.idat & GEOID_barcode_location_Red.idat (In the case of GEO data)**  
for example  
\- GSM2417753_200598790047_R01C01_Grn.idat  
\- GSM2417753_200598790047_R01C01_Red.idat
- ***inputtype: "idat" for idat files***  
- ***sample_names: a vector of sample names. If idat files are downloaded from GEO, sample_names should be "FALSE". Then, sample names are automatically downloaded and converted to GEO sample title baed on the file name.***


*lumiMethyNorm generates **Process_Result folder**: reports of the normalization, **processed_Mval.txt**: a matrix of M-value, and **sel_processed_Mval.txt**: a matrix of M-value which does not inculude low detection p-value probes. (default cut off is 0.01)*  

#### - idat files from GEO
```r
selDataMatrix <- getIdat2M(GEOid = GSE100825, version = "EPIC", sampleNames=FALSE)
```
- ***GEOid: a GEO accession ID***
- ***version: "27" or "450" or "EPC" or a vactor of multiple versions***
- ***sampleNames: If"FALSE", sample names are automatically downloaded and converted to GEO sample title baed on the file name.***  

For a GEO entry, you can directly download idat files and perform normalization and M-value computation by using getIdat2M function only with a GEO accession ID.  
If multiple version of idat files are registered in a GEO entiry, you cchoose a version(s) to be analyzed. If you choose multiple versions, the out put object is list of M-value data.frames.

  
### 3. motif database construction  
#### - Example 1: JASPER_CORE, Hsapiencs and Mmusclus
```r
library("MotifDb")
targetDB <- "JASPAR_CORE"
targetORG <- c("Hsapiens", "Mmusculus")
motifDB <- query(MotifDb, targetDB)        #extraction of motif list of "JASPER_CORE"
motifDB <- c(query(motifDB,targetORG[1]),query(motifDB,targetORG[2]))        #extraction of motifs of "Hsapiens" and "Mmusclus"
```
If you want analyze a specific motif select a motif. (ex. SPI1)
```r
targetTF <- "SPI1"
motifDB2 <- query(motifDB,targetTF)       #Extraction of motifs for target TF(s)
```
Finally, convert the motif list to list format.
```r
motifDBList <- as.list(motifDB)
```
#### - Example 2: IMAGE motif database
```r
motifDBList <- IMAGE_PWMlist
```
If you want analyze a specific motif select a motif. (ex. SPI1)
```r
targetTF <- "SPI1"
motifDBList <- IMAGE_PWMlist[grep(targetTF, names(IMAGE_PWMlist))]       #Extraction of motifs for target TF(s)
```

Motif list should be a list of PWMs of following format;  
  
   1     2    3    4    5    6    7    8    9  
A    0    0    0    0.1189189    0.1027027    0.2972973    0.28648649    0.10270270    0.04864865  
C    0    1    1    0.3837838    0.3081081    0.2378378    0.16216216    0.08648649    0.42162162  
G    1    0    0    0.2486486    0.3297297    0.3621622    0.49189189    0.74054054    0.42702703  
T    0    0    0    0.2486486    0.2594595    0.1027027    0.05945946    0.07027027    0.10270270  
  
 
### 4. Screening of enriched motifs  
```r
MotScr(infile=["sel_processed_Mval.txt" or selDataMatrix], motifDBList = [motif list( eg. motifDBList], cutoff = 2, p.cutoff = 0.05, outname="screening_result", ControlColnum=c(1,2), TreatmentColnum=c(3,4), MethylDemethyl="Demethyl", sampling=FALSE, version = "850")
```
- ***infile: M-value of infinium methylation array. text file or data frame object***  
- ***motifDBList: list format of PWM***  
- ***cutoff: cutoff velue of delta M***  
- ***outname: name of output file***  
- ***ControlColnum: A column of control data, such as unddiferentiated***  
- ***TreatmentColnum: A column of treatment data such as differentiated***  
- ***MethylDemethyl: "Methyl" (or "M") or "dimethyl (or"D").analysis target of differentially methylated probes***
- ***sampling: number of randomly sampled differentiall methylated probes. "FALSE" analyses all differentially methylated probes***  
- ***version: "450" or "EPIC" (or "850"). version of methylation array***  


*If you perform a comparison of multiple samples, identification of differentially methylated probes uses both Welch's t-test and M-value difference. For single sample comparison, it uses only M-value difference (delta M).*  

If you want to extract the motif of interest (e.g. SPI1):
```r
motif_names <- "SPI1"
motif <- motifDBList[grep(motif_names,names(motifDBList))]
MotScr(infile="sel_processed_Mval.txt", motifDBList = motif, cutoff = 2, p.cutoff = 0.001, outname="screening_result", ControlColnum=c(1,2), TreatmentColnum=c(3,4), MethylDemethyl="Demethyl", sampling=FALSE, version = "850")
```


 ### 5. output files
 - [outname_DMP_position.txt  
 DMP position
 - [outname]_mot_analysis_result.txt  
summary table
 - [outname]_plot.pdf  
 All histograms, enrichment score plots, fold-change plots, and p-value plots
 - [outname]_result.RData  
R data file
 - [outname]_sig_plots directory/[motif_name].pdf  
A directory containing enrichment score plots of significantly enriched motifs

### ***Scatter plot***
### 1. Reading of M-value data  
```r
infile <- "sel_processed_Mval.txt"
selDataMatrix <- read.table (infile)
```
### 2. Labels of the plot  
```r
main <- paste(colnames(selDataMatrix)[ControlColnum]," vs. ", colnames(selDataMatrix)[TreatmentColnum], sep="")
xlab <- paste("Control: ", colnames(selDataMatrix)[ControlColnum], sep="")
ylab <- paste("Treatment: ", colnames(selDataMatrix)[TreatmentColnum], sep="")
pdf ("Scatter_Plot.pdf")
```
### 3. Plot  
```r
scatterPlot(treatment_data=selDataMatrix[,8], control_data=selDataMatrix[,1], main=main, xlab=xlab, ylab=ylab, cutoff=2)
dev.off()
```

CHANGELOG.md
# Changelog
## [1.3] - 2018-08-07
### Added
- IMAGE motidf database (https://genome.cshlp.org/content/28/2/243.long)

### Fixed
- probeID2position_EPIC.R was integrated to ProbeID2positon.R and deleted.
- Conversion of duplicated probeIDs to position is available. (ProbeID2positon.R)
- dependecy of packages was fixed.
- Date was removed from output file.

## [1.4] - 2018-11-09
### Added
- sampling option
- Output of aignificantly enriched motif plots to a directory

### Fixed
- Numer of core to be used has been changed from 16 to 4.

## [1.5.0] - 2019-10-05
### Update
- Format of scripts was hanged to Roxygen format.

### Added
- When using the sampling option, indicate it on standerd out
- Output DMPs
- lumiMethyNorm support idat files as input
- lumiMethyNorm returns a selDataMatrix object
- MotScr support data.frame as input
- getIdat2M function: get, normalize, and M-value conputation from GEO ID

### Fixed
- Error with the case of DMR = 0/1

## [1.5.1] - 2019-12-27
### Fixed
- debug of probeID2position.R
- debug of seqExtract.R

## [1.5.2] - 2020-05-11
### Fixed
- add "force" option to lumiMethyNorm.R, which is inherited option from readEPIC

## [1.5.3] - 2020-05-23
### Updated
- add "set.seed" to stratSampling.R to fix the randomization

## [1.5.4] - 2020-06-11
### Updated
- singel data vs multiple data-compliant

## [1.5.5] - 2020-08-24
### Updated
- min.score paraneter of MachPWM is variable


R/MotScr.R
#' PWM overrepresentation analysis pipline for illumina methylation array
#' 
#' pipline to analyze enrichment of given motif PWMs for differentially methylated region based on Illumina infinium methylation array data.
#' 
#' @param infile input file: normalized M-value matrix
#' @param motifDBList PWM data list
#' @param cutoff cutoff for delta-M-value
#' @param p.cutoff cutoff for Welch's t-test
#' @param outname output files ID
#' @param ControlColnum column number for Control sample(s)
#' @param TreatmentColnum  column number for target sample(s)
#' @param MethylDemethyl direction of Ddifferentially methylated regions (methylated or demethylated)
#' @param version version of Infinium Methjylation array (450 or 850(orEPIC))
#' @param sampling If sampling number is indicated and DMP is more than sampling number, analysis is ran with randomly selected DMP. If FALSE, all DMPs are analyzed.
#' @param min.score The minimum score for counting a match. Can be given as a character string containing a percentage (e.g. "85%") of the highest possible score or as a single number. 
#' 
#' @import BSgenome.Hsapiens.UCSC.hg19
#' @importFrom utils read.table write.table
#' @importFrom stats runif na.omit
#' @importFrom grDevices pdf dev.off
#' 
#' @return analysis summary files and plots, Rdata
#' @keywords PWM, overrepresentation analysis
#' @export

MotScr <- function(infile,
	motifDBList, cutoff = 2,
    p.cutoff = 0.001,
    outname = "screening_result",
    ControlColnum, TreatmentColnum,
    MethylDemethyl = "Demethyl",
    version = "850",
    sampling = FALSE, min.score = "90%"
    ){
    if(length(motifDBList) == 0) stop("motifDBList not found")
    if(missing(infile)) stop("infile not found. data.frame or text file name.")

    if(is.character(infile) && grepl("txt$", infile)){
        #read the M-value text file
        cat("Reading M-value data...\n")
        selDataMatrix <- read.table(infile, sep = "\t")
    }

    cat("DMP identification...\n")
    DMP_IDs <- DmpId(selDataMatrix = selDataMatrix, ControlColnum = ControlColnum, TreatmentColnum = TreatmentColnum, p.cutoff = p.cutoff, cutoff = cutoff, MethylDemethyl = MethylDemethyl)
    if(length(DMP_IDs) == 0) stop("No DMP!!")

    if(!sampling == FALSE & length(DMP_IDs) >= sampling){
        ##fast option
        cat(paste0("Analysis will be run with ", sampling, "randomly selected DMPs."))
        cat("\n")
        DMP_IDs <- DMP_IDs[floor(runif(sampling, 1, length(DMP_IDs)))]
    }

    nDMP_IDs <- length(DMP_IDs)
    allProbe_IDs <- rownames(selDataMatrix)
    if(version == "450"){
        probe_annotation <- InfiniumDiffMetMotR::Methyl450anno
    } else if((version == "EPIC") || (version == "850")){
        probe_annotation <- InfiniumDiffMetMotR::EPICanno
    }

    target_position <- na.omit(probeID2position(probe_IDs = DMP_IDs, anno_info = probe_annotation)) #conversion of DMP IDs to position
    randomProbe_IDs <- stratSampling(target_IDs = DMP_IDs, anno_info = probe_annotation) #stratified sampling for negative control
    random_position <- na.omit(probeID2position(probe_IDs = randomProbe_IDs, anno_info = probe_annotation)) #conversion of NC probe IDs to position
    positionsList <- list("target" = target_position, "random" = random_position) #integrate DMP positoins and NC positions

    ##write DMP position
    DMPtOutfile <- paste0(outname, '_DMP_position.txt')
    out_target_position <- cbind(probeID = rownames(target_position), target_position)

    write.table(out_target_position, file = DMPtOutfile, quote = FALSE, row.names = FALSE, col.names = TRUE)

    ## sequence extraction
    ## Read human hg19 genomic sequence
    genome <- BSgenome.Hsapiens.UCSC.hg19
    cat("\nGetting the sequences...\n")
    seq_range <- c(-5000, 5000) #range from the CpG position to be extracted
    sequences <- lapply(positionsList, function(x){ seqExtract(positions = x, genome = genome, seq_range) })

    ##make a templrary outpput directory
    tempDir <- paste0(outname, "_temp")
    dir.create(tempDir)
    ##writing the sequences to splitted files
    seqs <- sequences$target
    target_all_filenames <- writeSplitSeq(seqs = seqs, split_num = 2500, tempDir = tempDir, output_file = "target")
    seqs <- sequences$random
    random_all_filenames <- writeSplitSeq(seqs = seqs, split_num = 2500, tempDir = tempDir, output_file = "random")
    rm(sequences)
    rm(seqs)
    invisible(replicate(3, gc()))

    cat("\n")
    cat(paste0("motif serch: Total ", length(motifDBList), " motifs"))
    cat("\n\tTarget regions...\n")
    ## ((multi-fasta file(multi-seqs) x motif) x [motif number])) x [multi-fasta file number]
    target_positionsList <- splitSeqMotDist(filenames = target_all_filenames, motif_list = motifDBList, min.score = min.score)
    ntarget_hits <- lapply(target_positionsList, function(x){ length(unlist(x)) })
    file.remove(target_all_filenames)
    cat("\tbackground regions...\n")
    random_positionsList <- splitSeqMotDist(filenames = random_all_filenames, motif_list = motifDBList, min.score = min.score)
    nrandom_hits <- lapply(random_positionsList, function(x){ length(unlist(x)) })
    file.remove(random_all_filenames)
    gc()

    cat("\nPlotting the results...\n")
    ##Plot output setting
    distPlotFile <- paste0(outname, '_plot.pdf') ##output file name setting
    pdf(distPlotFile)
    sigPlotDir <- paste0(outname, '_sig_plots')
    ifelse(!dir.exists(sigPlotDir), dir.create(sigPlotDir), FALSE) # make a directory for significantly enriched motifs
    parameter_matrix <- NULL
    All_motif_names <- names(motifDBList)

    ##Computation by each motif in motifDB
    for(j in 1:length(target_positionsList)){
        ##monitering of progress (standerd output of vidualizing motif name)
        indicator_2 <- paste("    ", "MOTIF   ", All_motif_names[j], ": ", j, "/", length(All_motif_names), "      ", date(), "\n", sep = "")
        cat(indicator_2)

        motif_name <- All_motif_names[j]
        target_mot_posi <- unlist(target_positionsList[[j]])
        ctrl_mot_posi <- unlist(random_positionsList[[j]])

        ## histogram plot
        cat("    Step 1 / Plot Histgram (Figure 1)...\n")
        motHist(target_mot_posi, ctrl_mot_posi, seq_range = seq_range, motif_name = motif_name)

        ## motif enrichment score plot
        cat("    Step 2 / Plot Enrichment Score (Figure 2)...\n")
        enrichment_scores <- enrichScoreDist(target_mot_posi, ctrl_mot_posi, seq_range = seq_range, motif_name = motif_name, nDMP_IDs = nDMP_IDs, outname = outname, plot_draw = TRUE)

        cat("    Step 3 / Creating a count.pvalue Matrix...\n")
        windowSize <- 100
        motif_counts_matrix <- countMatrix(target_mot_posi = target_mot_posi, ctrl_mot_posi = ctrl_mot_posi, seq_range = seq_range, windowSize = windowSize, slide = 50)

        if((length(target_mot_posi != 0)) && (length(ctrl_mot_posi != 0))){
            cat("    Step 4 / Visualization of FC (Figure 3)...\n")
            FCChangePlot(motif_name = motif_name, motif_counts_matrix = motif_counts_matrix)

            cat("    Step 5 / Poisson disribution model exact test (Figure 4)...\n")
            PoisTestPlot(motif_counts_matrix = motif_counts_matrix, motif_name = motif_name, seq_range = seq_range)
        }

        cat("    Step 6 / Extraction of significantly enriched ranges...\n")
        significant_ranges <- SigRange(motif_counts_matrix = motif_counts_matrix, cutoff = 0.00001, windowSize = windowSize)

        cat("    Step 7 / Significance test...\n")
        parameters <- enrichTest(significant_ranges = significant_ranges, motif_counts_matrix = motif_counts_matrix, seq_range = seq_range, enrichment_scores = enrichment_scores, windowSize = windowSize)
        ifelse(!parameters["peak.Test"] == "significant" & file.exists(paste(outname, '_sig_plots/', motif_name, '.pdf', sep = "")),
        file.remove(paste0(outname, '_sig_plots/', motif_name, '.pdf')), FALSE) ##remove the sig_plot_file if enrichment is not significant
        parameter_matrix <- rbind(parameter_matrix, parameters)
    }
    dev.off()

    cat("\nResult table writing...\n")
    finOut <- cbind(names(motifDBList), ntarget_hits, nrandom_hits, parameter_matrix)
    ##out put file name setting
    ResultOut <- paste0(outname, '_mot_analysis_result.txt')
    ResultOutR <- paste0(outname, '_result.RData')
    write.table(finOut, file = ResultOut, sep = "\t", quote = F, row.names = F)
    save(nDMP_IDs, seq_range, outname, All_motif_names, positionsList, target_positionsList, random_positionsList, file = ResultOutR)
    file.remove(tempDir)
    cat("Completed!!\n")
}
